docs: Cumulocity certificate authority documentation (#3590)
thin-edge/thin-edge.io: docs/src/references/certificate-management.md

---
title: Certificate Management
tags: [Reference, Certificate, Security]
sidebar_position: 1
description: Certificate Management reference guide
---

import UserContext from '@site/src/components/UserContext';
import UserContextForm from '@site/src/components/UserContextForm';

:::tip
#### User Context {#user-context}

You can customize the documentation and commands shown on this page by providing
relevant settings which will be reflected in the instructions. It makes it even
easier to explore and use %%te%%.

<UserContextForm settings="C8Y_PROFILE_NAME,C8Y_PROFILE_URL,C8Y_URL,DEVICE_ID" />

The user context will be persisted in your web browser's local storage.
:::

> ## Integration with Cumulocity CA {#cumulocity-certificate-authority}

Author: A one-line description of the feature is missing.

Reviewer: hm, this is getting a bit difficult to describe as the page is definitely 95% c8y focused, so I'm questioning the overall title of the page now.

Author: It's currently c8y focused because only that CA is supported for now. Somehow this is what the intro can tell.

I will come with a proposal.

Author: See https://github.com/thin-edge/thin-edge.io/pull/3590/commits/718cf509ff4da9aac707a429ac12ddec086c09d2

:::

## Certificate renewal

Once the %%te%% device is connected to Cumulocity, its certificate can be renewed by using the existing connection to Cumulocity. If the device is using SystemD as its init system, then the certificate certificate renewal process will be automatic, and it will be renewed before the certificate expires (the default is 30 days before expiration, but the value can be configured via the tedge configuration).

The Cumulocity certificate-authority feature provides a new API endpoint for devices to request a new certificate provided the device can still communicate with the tenant it is requesting a certificate from. The certificate renewal is authenticated by way of a JWT which is requested over the MQTT connection using the current certificate. If for any reason the certificate expires, and the device can no longer communicate with Cumulocity, then the device will need to be registered again.

To ensure a smooth transition to the new certificate, the certificate renewal process is done in two distinct steps as described below:

> 1. Request a new certificate authenticating with the current certificate. The newly issued certificate is downloaded to the device, however it will not overwrite the current certificate. Instead the newly issued certificate will be saved next next to the current certificate and be given the `.new` filename suffix.

Author: ```suggestion
1. Request a new certificate authenticating with the current certificate. The newly issued certificate is downloaded to the device, however it will not overwrite the current certificate. Instead the newly issued certificate will be saved next to the current certificate and be given the `.new` filename suffix.
```

Reviewer: Fixed in f3a099ff38ed6949b893fdb6e704fa013c82df16

2. Verify the connection to Cumulocity using the new certificate candidate, and if the connection to Cumulocity is successful, then the new certificate candidate will overwrite the previous certificate. If the connection is not successful, then the existing certificate is not touched, and the certificate candidate will also remain untouched.

For additional options on how the certificate registration and renewal can be configured (e.g. how many days before the certificate expires should the certificate be renewed), check out the [Configuration](#configuration) section.

### Automated certificate renewal (systemd)

> If you're using a device with SystemD, connected to Cumulocity and have the , then the certificate renewal service will be installed, and it will take care of renewing the device certificate automatically.

Author: Missing words

```suggestion
If you're using a device with SystemD, connected to Cumulocity and have the <??? Missing words ???>, then the certificate renewal service will be installed, and it will take care of renewing the device certificate automatically.
```

Reviewer: Fixed in f3a099ff38ed6949b893fdb6e704fa013c82df16 (I just removed the 3rd condition which would be the "certificate authority" feature being enabled...but this can be implied otherwise it makes the sense harder to read)

The automated certificate renewal is implemented using SystemD and the components and their role are listed below:

|SystemD Unit|Description|
|------------|-----------|
|tedge-cert-renewer@c8y.timer|Timer which runs periodically to trigger the `tedge-cert-renewer@c8y` service. By default it runs every hour on the hour (with built-in jitter).|

> |tedge-cert-renewer@c8y.service|Service which performance the renewal if the certificate is about to expire as defined by in the %%te%% configuration|

Author: ```suggestion
|tedge-cert-renewer@c8y.service|Service which performs the renewal if the certificate is about to expire as defined in the %%te%% configuration|
```

Reviewer: Fixed in f3a099ff38ed6949b893fdb6e704fa013c82df16
